src/Functions/multiIf.cpp

```diff
@@ -374,17 +378,54 @@ class FunctionMultiIf final : public FunctionIfBase
     }
 
     template <typename T, typename S>
-    static void executeInstructionsColumnar(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res)
+    static void executeInstructionsColumnar(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res, const MutableColumnPtr & null_map, bool nullable)
     {
         PaddedPODArray<S> inserts(rows, static_cast<S>(instructions.size()));
         calculateInserts(instructions, rows, inserts);
 
         PaddedPODArray<T> & res_data = assert_cast<ColumnVector<T> &>(*res).getData();
-        for (size_t row_i = 0; row_i < rows; ++row_i)
+        if (!nullable)
         {
-            auto & instruction = instructions[inserts[row_i]];
-            auto ref = instruction.source->getDataAt(row_i);
-            res_data[row_i] = *reinterpret_cast<const T*>(ref.data);
+            for (size_t row_i = 0; row_i < rows; ++row_i)
+            {
+                auto & instruction = instructions[inserts[row_i]];
+                auto ref = instruction.source->getDataAt(row_i);
+                res_data[row_i] = *reinterpret_cast<const T*>(ref.data);
+            }
+        }
+        else
+        {
+            PaddedPODArray<UInt8> & null_map_data = assert_cast<ColumnUInt8 &>(*null_map).getData();
+            std::vector<const T*> data_cols(instructions.size());
+            std::vector<const UInt8 *> null_map_cols(instructions.size());
+            for (size_t i = 0; i < instructions.size(); ++i)
+            {
+                if (instructions[i].source->isNullable())
+                {
+                    const ColumnNullable * nullable_col;
+                    if (!instructions[i].source_is_constant)
+                        nullable_col = assert_cast<const ColumnNullable *>(instructions[i].source.get());
+                    else
+                    {
+                        const ColumnPtr data_column = assert_cast<const ColumnConst &>(*instructions[i].source).getDataColumnPtr();
+                        nullable_col = assert_cast<const ColumnNullable *>(data_column.get());
+                    }
+                    null_map_cols[i] = assert_cast<const ColumnUInt8 &>(*nullable_col->getNullMapColumnPtr()).getData().data();
+                    data_cols[i] = assert_cast<const ColumnVector<T> &>(*nullable_col->getNestedColumnPtr()).getData().data();
+                }
+                else
+                {
+                    null_map_cols[i] = ColumnUInt8::create(rows)->getData().data();
```

Comment:
We should also probably reuse a single `ColumnUInt8` for different columns

Reply:
done @evillique

Reply:
Well it looks like if the first column was nullable then we would use some nullable filter on a non-nullable column

Reply:
Fixed . @evillique